Add schema v4.4

spbnick opened the pull request:
Add schema v4.4 with support for describing issue presence transitions.

The transitions point out the range of revisions along the change history, where an issue has appeared or disappeared from the code. Note that although there are no explicit revision objects in the I/O schema they exist as aggregated objects in the ORM and the dashboards.

Transitions are generally expected to be substantiated by issue incidents (which can be inferred from the revision and the issue linked from the transition). However, that's not really necessary, as long as the submitter makes it clear what happened in the transition and/or issue description (or perhaps in the linked report), and makes sure it really did.

Transitions can be "updated" - new versions of them submitted, when e.g. they're located more precisely (the revision range was narrowed), or for example they need to point to another issue.

spbnick commented:
OK, here's what I came up with so far, @hardboprobot.

I decided to go with a much lighter coupling between regression/recoveries and actual test artifacts, making the presence of the latter optional, and instead pointing to the boundary revisions directly. This is an optimization after all, we can as well make it easier to query.

I don't have any more time for trying to express your stats with this, but I should be able to do that tomorrow.

Please post the cases you think are the most difficult and/or important here, if you can. And I'll try to work them out.

Thanks!

tales-aparecida commented:
In reply to https://github.com/kernelci/kcidb-io/pull/78#issuecomment-2124349303

> I think the creation of issues depends only on the info provided by the test results, and thus it can be platform and CI-agnostic -> it can be done by an independent and standalone process (we can work on a tool for this), but identification of transitions is probably easier to do if we factor in the data origin, ie. I think it's easier to identify a transition looking at localized results in a per-CI basis.

Based on my experience with [CKI](https://cki-project.org), where most Issues can be identified by a set of regexes over the logs generated by a single test, I can see how this _might_ be true...

Just so I can understand this better, in that scenario you are proposing, we would only _automatically_ create issues when processing test runs about trees, right? Patches that are still being reviewed have the potential to introduce issues, but they don't need to be tracked until they land IMO.

I'm really asking you this because, _from my perspective_, the CKI workflow cares more about tests runs on patches (MRs), so we're still relying on human judgement to categorize whether the patch caused the FAIL or not, e.g. comparing the context of the failure with the lines of code that changed; comparing the outcome of the same tests in builds from other architectures and configs, that might've ran in different hosts; manually checking ticket trackers for other trees; and even asking people if they had seen that before. Once we are sure the patch is not responsible for the FAIL, we create a new issue (which points to a ticket tracker with someone assigned to fix it ASAP), otherwise the developer should update their code and start a new test run.

tales-aparecida commented:
Now that I see the solidification of the idea of `transition`, I'm struggling to see how we'd go about dealing with flaky bugs in it. I would like to think that they would just be ignored until someone is sure the bugs are gone, but I can also see them being "toggled" with a pair of on-and-off transitions whenever they show up and inevitably disappear, if we're not careful.

I would like to see some examples or mock ups of the data that would be stored using this if we have time to discuss this further before merging.

spbnick commented on kcidb_io/schema/v04_04.py:
```diff
@@ -0,0 +1,1129 @@
+                            "The Git repository branch from which the commit "
+                            "with the base source code was checked out."
+                    },
+                    "git_commit_generation": {
```
Yep, this won't support that.

spbnick commented:
> According to the schema descriptions, issues, incidents and transitions are meant to be generated in the source CI systems and then submitted to KCIDB. Have you though about the potential pros and cons of this approach vs having the CI systems submitting only the results and having KCIDB generate issues, incidents and transitions from them?

My aim is to support both cases, as needed. Have KCIDB do its own triaging and reporting on submitted results, as much as it can. However, if a CI system is equipped with more advanced machinery for detecting particular issues or regressions, let them do it until KCIDB catches up.

For our particular target of having your report generation work in KCIDB, my aim is to have the detection of conditions and generation of the appropriate objects just *hardcoded* inside KCIDB for the start. That would make sure we have enough data reported in Maestro results for processing, and verify that we can express the outputs.

Next step would be adding condition fields to issues, and building a generic machinery for doing the same thing, but executing the matching expressed in the database, corresponding to what you need, and what we were hardcoding before.

spbnick commented:
> Now that I see the solidification of the idea of transition, I'm struggling to see how we'd go about dealing with flaky bugs in it. I would like to think that they would just be ignored until someone is sure the bugs are gone, but I can also see them being "toggled" with a pair of on-and-off transitions whenever they show up and inevitably disappear, if we're not careful.

Yes, this is a very valid concern. It is orthogonal to the concept of regressions/transitions, though. It's a question of how well we can trust a test result. We would have to have a layer which employs a measure of stability and statistic analysis to make this more robust. I don't know how to do this exactly yet. But small steps. Let's figure this out first.

> I would like to see some examples or mock ups of the data that would be stored using this if we have time to discuss this further before merging.

Yes, that's what I would like to do. First with the reports that are sent by the bot to the #general channel already. And we're far from merging this, in addition to that too, as we'll have to run this by submitters, after we agree on it, and before merging.

spbnick commented:
>  * Create higher-level "regression" objects after applying a band-pass filter to weed out unwanted results (flaky tests, false positives, etc)
> * Report flaky tests to lab maintainers and test authors, depending on the cause of the flakiness

Yeaaaah, that sounds good. I wonder if we should allow incidents to refer to transitions :thinking:

spbnick commented:
Thank you for another detailed and thoughtful response, @hardboprobot!

> In order for the high-level approach to be practical, though, I think we first need a way to characterize and identify issues, which is what I'm working on now.

I think we're looking at the problem from two somewhat different angles (which is great!). I'm coming from the environment where we have an army of QE maintaining specific tests an handling failures by manually creating issues which match the corresponding results as closely as possible. They know the test, know the environment and can do that.

These issues can generally be applied to the full swath of trees we're testing. In addition they enjoy the limitation of mostly testing the latest commits (although they can be to quite old branches). Whenever there's a tree-specific problem, it's possible to add a corresponding condition to the issue to limit detection. There's constant collaboration, and negotiation on and about the issue conditions.

If I understand you right, you would like to have automatic **identification**, and **scoping** of issues. This is great and I fully support these efforts. And I think that we would still need to have the same **base** schema to implement the low level, as you say. Plus, as we already agreed before, it would be great if a human can interfere and correct the automatically-generated issue description in any case. So, I think we can get on with implementing what we have so far (or at least attempting to). This will help us figure out the foundation for the automatic detection and having the former more defined will help us think about the latter.

> But let's take it bit by bit:
> 
> > In the context of the whole database, ACTIVE would mean that there's at least one tree/repo/branch which don't have a confirmed (tested) fix. That won't be very useful, probably, as it takes a long time to propagate (and test) the fix, and will never happen for removed/abandoned branches. The tree/repo/branch-context view would contain a more useful status. Still it needs to be here for completeness.
> 
> Here we hit the problem I mentioned in [KCIDB design and development ideas](https://docs.google.com/document/d/1xbFcXOdpS4umrE1oYy2JXuDYsgnk1ROGLQmByTxoi7s/edit#heading=h.swy9beix2hp8), that in order to do this across different trees/branches the issues need to be sufficiently specific. Something like what you proposed there: "Test storage.WriteZeroPerf failing" may be useful for a human to know what the issue is about but isn't specific enough to tell two issues apart. In this scenario, the same test may be failing in two different branches/repos for different reasons.

Yes, that "Test storage.WriteZeroPerf failing" is absolutely the simplest, wildcard case. And it's coming from the requests from maintainers to only send them "new" failures. It's crude, but even just that could make KCIDB reports and dashboards more useful, and potentially attract new users, and to give us feedback. And a simple case is good to start the implementation with.

> I'd say for now we can either:
> 
>     * assume we have a way to profile issues (I'm working on that) and move on.

I think this is a good idea.

>     * simplify the detection of FAIL->PASS transitions by restricting the search by the tree/branch (what KernelCI currently does)

Let's also keep this in mind and see how things go with the implementation and with the actual data, and tackle things as they come.

> > > Introduced in: 2024-05-12T23:09:31.805000
> 
> > For the whole database context this would simply be the minimum start_time across all tests referenced by incidents marking the issue as present. We don't need to reference transitions for that.
> 
> I'm not completely sure of that. Let's suppose an example: a test failed last month causing a transition (PASS->FAIL) identified by a specific issue, and an incident (present). Two weeks ago the test was detected to pass, so we get a new transition (FAIL->PASS) and another incident (absent). Btw, please, correct me if I'm wrong about the use of incidents, I think I see some overlapping of functionality now between concepts but I'm not sure. Now, the same test fails again today, the culprit is a different commit or whatever other condition, but the issue seems to be the same. So we get another transition (PASS->FAIL) and another incident (present) pointing to the same issue.
> 
> If we search for the minimum start_time across tests referenced by incidents marking the issue as present we'll get the result from last month, but that's unrelated to the one that happened today.
> 
> This example happens all the time in unstable tests in KernelCI, due to poorly programmed tests, setup problems, faulty hardware, etc. Only the frequency is much higher (multiple PASS->FAIL->PASS transitions in the same week).

Ah, of course. Just the timestamp didn't give me enough information on what it is about. We can look for the earliest commit with the issue instead here.

> > > * Logs:
> > >   
> > >   * lava_log
> 
> > Um, which log would that be? Earliest? Latest?
> 
> This should be the log from the earliest failure.

Right. Shouldn't be a problem.

> For the rest of the attributes, you mention how to get the results applied to the whole database. I think before this we should start from the primitive data as in the examples, where we have a very zoomed-in view of a test failure. And then, when we have a way to characterize the failure we can zoom-out and search for the same issue across the whole DB.

Well, sure. Although having one more dashboard for the overall database doesn't hurt anyone and is easy to implement in Grafana. Notifications is another thing.

> > Overall this seems doable from the POV of the dashboard and SQL. The incident objects really make things simpler and allow the dashboard to be completely unaware of how issues are matched. I have conflicting feelings about transitions. In many ways they feel rudimentary, as a lot of stuff can be done with just incidents, but they do provide the shortcut to revisions. Let's get on with the implementation and things will get clearer.
> 
> Maybe another way to interpret the roles of transitions and incidents is this separation between low-level and high-level processing that I'm proposing rather than trying to conflate them into the same solution space. That is, transitions may be meant to provide the solution to one problem (rudimentary detection of breaking points) and incidents provide a solution to a different problem (relationship between test failures and issues). Linking the information provided by these two may be a way to bring low-level information to the high-level problem.

I agree.

> The next three weeks I'll be available to work on anything related to this, so let me know if I can help or assist you in any way.

Great! I'll start updating the database drivers ASAP. Just need to finish jq.py rebasing. Then we would be able to start reimplementing the issue detection in KCIDB, in a separate deployment. That's where I will need your help. When we see the things are going the right way, we'll send the schema proposal to the submitters.

In parallel, if you feel we'll need to look at the logs for our issue detection soon, we'll need to improve artifact caching, to increase speed, and save on incoming traffic costs. At the moment the caching logic is rather trivial, and also it only downloads every 256th file (to save costs, again, for the testing period). We'll need to figure out storage policy so that data is retired to cheaper storage automatically, and make sure we cache/archive all the files we need. You could also help with that. Unless we find someone else.

spbnick commented:
I finished rebasing our fork of jq.py, and will be able to start the driver work on Monday.

spbnick commented:
One thing this schema doesn't specify is how to "remove" a transition. There needs to be a certain value of a field or combination of fields that say "this transition is invalid". Missing fields mean "unknown", so that is not available.

spbnick commented:
Perhaps we could permit empty `issue_id` (distinct from missing `issue_id`)?
Speaking of which, I'm wondering how to make an issue itself invalid.

spbnick commented:
> That a good plan, ping me if you have a clear idea of where to start or have already some tasks that you want to offload. Or we can have a quick meeting or slack chat about it.

Great! Let me finish the basic database support and post it for your review. And then we take it from there.

> Sorry, I may have forgotten part of the context, but I don't recall talking about this concept of a "removed" or "invalid" transition. Which case do you want to model with that?

We didn't discuss it, but it's very likely that a transition we "find" might turn out to be wrong and we might need to "remove" it. Same with issues. We cannot actually delete stuff from KCIDB, so we need a way to invalidate issues, incidents, and transitions.

spbnick commented:
Also, do you think we'll need to look inside log files soon? If yes, then we need to improve the artifact cache.

spbnick commented:
Ah, incidents don't have versions, and they don't need to be invalidated, but issues do.

spbnick commented:
Getting there. DB support just passed all local tests. Tests against a deployment are next.

spbnick commented:
Alright, preliminary support for this schema is up in a PR: https://github.com/kernelci/kcidb/pull/529

@hardboprobot, please review!

spbnick commented:
OK, this comment **is** long, so I gave up trying to address specific points separately, and will just dump everything I have to say about it, so far.

I think it will be more productive to exchange shorter comments more often in the future, and perhaps have meetings to just talk over stuff.

The "issue extractor" (my name for the automatic issue creator you're working on for now) can only know if the issue in the result already exists by parsing the result first.

The question is, whether, once extracted, it can describe the issue in some other way (at all), and whether that other way would allow something else to identify it faster. E.g. regular expressions or shell globs. If not, or if the difference is not significant, then we don't have to do that, and can simply feed *every* result to the extractor, and let it maintain its own issue IDs and versions. The issues it creates wouldn't be used by a generic triager to find more incidents, but only the extractor itself will be submitting them. In this case it would be similar to just another CI system submitting issues and incidents.

If it can generate something else to match the issue with, e.g. a regex or (better) a glob, which would run faster, then the only thing the issue extractor should do is (re)submit the issues. The triager (the thing which matches generic issues to results) should submit the incidents, and take care of not submitting more than necessary.

Now, we can't possibly triage **every** result with **every** issue, we'll have to pick our fights, prioritize which issues are most important to triage, and have a cut off time before generating the first round of notifications about the results. We can continue (e.g. opportunistically) triaging afterwards, and uncover more correlation, but we gotta call "triaging results are good enough" at some point.

> Problem: a transition is only related to an issue in the schema, but they should be also linked to the test runs that define it (at least the most recent one, the one that creates the transition)

The test runs can be found by querying tests linked to issue's incidents (and the transition's revision) and picking the earliest result. Also, there's no guarantee, that it will be the earliest test that created the transition. They can arrive in any order. And it's really not that important if we get exactly the earliest test for that, or exactly the test that triggered the creation of the transition, it's the issue and the revision identification that matters. We can always sort the incident tests by `start_time` for a particular revision.

Then, there's really no answer to the question "in which branch an issue first appeared", in general, given enough time has passed to disseminate the commit among the trees.

However, we can find where we checked out that commit from first (time-wise), which could be good enough. And the solution improves with increased sample rate and sample width (checking more branches more often).

In theory, we could refer to the "origin" test/build result which "produced" the issue from the issue itself. Like have a special field for that. However, if the issue extractor submits multiple issues with this field different, the outcome will be indeterministic and the whole point of having this data will be missed.

That could be used in manually-created issues, though.

> Possible solution: link "transitions" to the related test runs / builds.
> Substitute revision_before_* and revision_after_* with test_id_before / test_id_after and build_id_before / build_id_after. But this is cumbersome because a transition is either for a test or for a build, but not for both. Possible solution to this: split transitions into "Test transitions" and "Build transitions" and make the reference ids not null.

I started designing the schema with this idea. However, I then changed it to the present schema. First, this would need considerably more joins for the major use-cases of showing which issues a branch has active, and showing the issue history. Especially for tests. Second, sometimes we might have no test results for particular revisions, but nevertheless might know that the issue has (dis)appeared there, due to out-of-system testing, or just a human investigation. If we insist on requiring incidents we would need to either fake them, or wait for a test to reproduce them, which might take a while for difficult-to-reproduce issues, and all this time the transition would be either pointing to a wrong commit or to nothing at all. It also could be that there's no test reproducing a certain issue, but it could be good to still record where it's introduced/fixed, thus augmenting a bug tracker.

Also, we had some discussion on how to integrate KCIDB with regzbot data this week, and we can surface its information as issues and transitions *without* incidents, which would already help us display them, where necessary. The regzbot-collected issues have a good reputation in the community, and it would be good to have them in the dashboards.

In general I find coupling gaps like this useful, as it gives us more flexibility and the system more viability.

Regarding linking both builds and tests from incidents, in theory (a bit far-fetched, but still) it's possible for an issue to manifest both in a build and in a test. E.g. a build can warn us about something, and then a test can actually trip on it. I don't think we need to worry much about both references being present there, and it makes things simpler. I.e. by keeping a single type of transitions. Eventually we might want to unify build and test results, like you guys did in Maestro, and then the problem (such as it is) will go away.

> We already have a relationship that links test and build objects with issues: incidents. So maybe we can encode the information of a transition in an incident and discard the "Transition" objects altogether.

> With these new fields this should provide the same information as the transition entity, since a transition is now equivalent to an incident with transition == true.

> Additionally, this removes a small redundancy that we had with the incidents.present and transition.appearance fields: they meant more or less the same thing, but applied to different objects.

I get where you're coming from. And yes, it's good to not have redundancy and unnecessary object types. We don't really have to have "transitions", and we can run SQL queries to figure out where an issue first or last had a positive incident, without them. However we'd like to have them as an optimization, and to express issue (dis)appearance without actual incidents. And optimization results often have redundancy, unfortunately.

I think putting previous build/test ID into incidents makes them accommodate two layers, which can be separate and implemented by different parties. Not many CI systems would be able to provide those fields, we cannot edit submitted CI data (we would have to use our own origin), so we cannot add those fields for them ourselves. OTOH, we can definitely build transition detection based on pure incidents they submit. As well as be able to cleanly separate issue triaging from regression detection in our own processing. Additionally the previous build/test ID in an incident could very quickly become outdated. E.g. once we do bisections and add more builds/tests between those the incident knows about, and we don't have a way to update incidents by themselves (and I don't think we should), only when issues are updated.

Anyway, I spent long enough trying to make this response coherent, while getting distracted by all matter of things :grin:

Let's exchange our thoughts and ideas more often, so we don't have to struggle replying, and please correct me if I got anything wrong, or I don't make sense :grin:

spbnick commented:
I spent some time thinking about "git commit generation" which we included here, and I think that we don't really need it. I think we should just use the checkout's start_time. The commit generation that we wanted to introduce would be essentially that anyway. I investigated using git rev-list --count HEAD, but it would still not allow us to compare history for force-push branches. For that, timestamp works better, actually. So I think we can drop it from the design, and work with start_time until and if we get the graph database.

spbnick commented:
**But**, we would need a field specifying that this was the **tip** of the branch at the checkout time. Because we can have checkouts for previous commits too.
